File: sdk/python/aragora_sdk/namespaces/checkpoints.py
"""
Checkpoints Namespace API.

Provides debate checkpoint management for pause, resume, and intervention.

Features:
- Debate checkpoint creation and management
- Pause and resume debate functionality
- Human intervention support
- Knowledge Mound checkpoint management
"""

from __future__ import annotations

import warnings
from typing import TYPE_CHECKING, Any, Literal

if TYPE_CHECKING:
    from ..client import AragoraAsyncClient, AragoraClient

CheckpointStatus = Literal["active", "resumed", "expired"]


def _warn_deprecated(message: str) -> None:
    """Emit a runtime DeprecationWarning for a dead or drifted SDK method."""
    warnings.warn(message, DeprecationWarning, stacklevel=3)


class CheckpointsAPI:
    """
    Synchronous Checkpoints API.

    Provides methods for checkpoint management:
    - List and manage debate checkpoints
    - Pause and resume debates
    - Perform human interventions
    - Knowledge Mound checkpoint operations

    Example:
        >>> client = AragoraClient(base_url="https://api.aragora.ai")
        >>> # List active checkpoints
        >>> checkpoints = client.checkpoints.list()
        >>> # Pause a running debate
        >>> checkpoint = client.checkpoints.pause_debate("debate_123")
        >>> # Resume later
        >>> result = client.checkpoints.resume(checkpoint['id'])
    """

    def __init__(self, client: AragoraClient) -> None:
        self._client = client

    # =========================================================================
    # Debate Checkpoints
    # =========================================================================

    def list(
        self,
        limit: int | None = None,
        offset: int | None = None,
    ) -> dict[str, Any]:
        """
        List all checkpoints.

        Args:
            limit: Maximum number of checkpoints to return.
            offset: Pagination offset.

        Returns:
            Dict with list of checkpoints and total count.
        """
        params: dict[str, Any] = {}
        if limit is not None:
            params["limit"] = limit
        if offset is not None:
            params["offset"] = offset

        return self._client._request(
            "GET",
            "/api/v1/checkpoints",
            params=params if params else None,
        )

    def list_resumable(self) -> dict[str, Any]:
        """
        Get resumable debates with active checkpoints.

        Returns:
            Dict with list of resumable debates, each containing
            debate_id, checkpoint_id, task, round, and paused_at.
        """
        return self._client._request("GET", "/api/v1/checkpoints/resumable")

    def get(self, checkpoint_id: str) -> dict[str, Any]:
        """
        Get a specific checkpoint.

        Args:
            checkpoint_id: The checkpoint identifier.

        Returns:
            Checkpoint details including id, debate_id, status, round,
            created_at, expires_at, and metadata.
        """
        return self._client._request("GET", f"/api/v1/checkpoints/{checkpoint_id}")

    def resume(self, checkpoint_id: str) -> dict[str, Any]:
        """
        Resume a debate from a checkpoint.

        Args:
            checkpoint_id: The checkpoint identifier.

        Returns:
            Dict with debate_id and resumed status.
        """
        return self._client._request("POST", f"/api/v1/checkpoints/{checkpoint_id}/resume")

    def delete(self, checkpoint_id: str) -> dict[str, Any]:
        """
        Delete a checkpoint.

        Args:
            checkpoint_id: The checkpoint identifier.

        Returns:
            Dict confirming deletion.
        """
        return self._client._request("DELETE", f"/api/v1/checkpoints/{checkpoint_id}")

    def intervene(
        self,
        checkpoint_id: str,
        action: str,
        message: str | None = None,
        config: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """
        Perform an intervention on a checkpointed debate.

        Allows human operators to modify debate parameters or inject
        information before resuming.

        Args:
            checkpoint_id: The checkpoint identifier.
            action: The intervention action (e.g., 'inject_context', 'modify_agents').
            message: Optional message to inject.
            config: Optional configuration changes.

        Returns:
            Dict with success status and optional message.
        """
        data: dict[str, Any] = {"action": action}
        if message is not None:
            data["message"] = message
        if config is not None:
            data["config"] = config

        return self._client._request(
            "POST",
            f"/api/v1/checkpoints/{checkpoint_id}/intervention",
            json=data,
        )

    # =========================================================================
    # Debate-Specific Checkpoint Operations
    # =========================================================================

    def pause_debate(self, debate_id: str) -> dict[str, Any]:
        """
        Pause a debate and create a checkpoint.

        Args:
            debate_id: The debate identifier.

        Returns:
            The created checkpoint with pause state.
        """
        return self._client._request("POST", f"/api/v1/debates/{debate_id}/pause")

    # =========================================================================
    # Knowledge Mound Checkpoints
    # =========================================================================

    def list_km(self, limit: int | None = None) -> dict[str, Any]:
        """
        List Knowledge Mound checkpoints.

        Args:
            limit: Maximum number of checkpoints to return.

        Returns:
            Dict with list of KM checkpoints.
        """
        params: dict[str, Any] = {}
        if limit is not None:
            params["limit"] = limit

        return self._client._request(
            "GET",
            "/api/v1/km/checkpoints",
            params=params if params else None,
        )

    def create_km(
        self,
        name: str,
        workspace_id: str | None = None,
    ) -> dict[str, Any]:
        """
        Create a Knowledge Mound checkpoint.

        Args:
            name: Name for the checkpoint.
            workspace_id: Optional workspace ID.

        Returns:
            The created KM checkpoint with name, created_at, size_bytes, etc.
        """
        data: dict[str, Any] = {"name": name}
        if workspace_id is not None:
            data["workspace_id"] = workspace_id

        return self._client._request("POST", "/api/v1/km/checkpoints", json=data)

    def get_km(self, name: str) -> dict[str, Any]:
        """
        Get a Knowledge Mound checkpoint.

        Args:
            name: The checkpoint name.

        Returns:
            KM checkpoint details including name, workspace_id, created_at,
            size_bytes, node_count, and metadata.
        """
        return self._client._request("GET", f"/api/v1/km/checkpoints/{name}")

    def compare_km(self, name: str, compare_to: str) -> dict[str, Any]:
        """
        Compare two Knowledge Mound checkpoints.

        Args:
            name: The first checkpoint name.
            compare_to: The second checkpoint name to compare against.

        Returns:
            Dict with checkpoint_a, checkpoint_b, additions, deletions,
            modifications, and details.
        """
        return self._client._request(
            "GET",
            f"/api/v1/km/checkpoints/{name}/compare",
            params={"compare_to": compare_to},
        )

    def compare_km_checkpoints(self, checkpoint_a: str, checkpoint_b: str) -> dict[str, Any]:
        """
        Compare two named Knowledge Mound checkpoints directly.

        Uses the documented POST /api/v1/km/checkpoints/compare contract
        (body: checkpoint_a, checkpoint_b).

        Args:
            checkpoint_a: First checkpoint name.
            checkpoint_b: Second checkpoint name.

        Returns:
            Dict with checkpoint_a, checkpoint_b, additions, deletions,
            modifications, and details.
        """
        return self._client._request(
            "POST",
            "/api/v1/km/checkpoints/compare",
            json={"checkpoint_a": checkpoint_a, "checkpoint_b": checkpoint_b},
        )

    def restore_km(self, name: str) -> dict[str, Any]:
        """
        Restore a Knowledge Mound checkpoint.

        Args:
            name: The checkpoint name to restore.

        Returns:
            Dict with restored status.
        """
        return self._client._request("POST", f"/api/v1/km/checkpoints/{name}/restore")

    def delete_km(self, name: str) -> dict[str, Any]:
        """
        Delete a Knowledge Mound checkpoint.

        Args:
            name: The checkpoint name to delete.

        Returns:
            Dict confirming deletion.
        """
        return self._client._request("DELETE", f"/api/v1/km/checkpoints/{name}")


class AsyncCheckpointsAPI:
    """
    Asynchronous Checkpoints API.

    Example:
        >>> async with AragoraAsyncClient(base_url="https://api.aragora.ai") as client:
        ...     # Pause a debate
        ...     checkpoint = await client.checkpoints.pause_debate("debate_123")
        ...     # Resume later
        ...     result = await client.checkpoints.resume(checkpoint['id'])
    """

    def __init__(self, client: AragoraAsyncClient) -> None:
        self._client = client

    # =========================================================================
    # Debate Checkpoints
    # =========================================================================

    async def list(
        self,
        limit: int | None = None,
        offset: int | None = None,
    ) -> dict[str, Any]:
        """List all checkpoints."""
        params: dict[str, Any] = {}
        if limit is not None:
            params["limit"] = limit
        if offset is not None:
            params["offset"] = offset

        return await self._client._request(
            "GET",
            "/api/v1/checkpoints",
            params=params if params else None,
        )

    async def list_resumable(self) -> dict[str, Any]:
        """Get resumable debates with active checkpoints."""
        return await self._client._request("GET", "/api/v1/checkpoints/resumable")

    async def get(self, checkpoint_id: str) -> dict[str, Any]:
        """Get a specific checkpoint."""
        return await self._client._request("GET", f"/api/v1/checkpoints/{checkpoint_id}")

    async def resume(self, checkpoint_id: str) -> dict[str, Any]:
        """Resume a debate from a checkpoint."""
        return await self._client._request("POST", f"/api/v1/checkpoints/{checkpoint_id}/resume")

    async def delete(self, checkpoint_id: str) -> dict[str, Any]:
        """Delete a checkpoint."""
        return await self._client._request("DELETE", f"/api/v1/checkpoints/{checkpoint_id}")

    async def intervene(
        self,
        checkpoint_id: str,
        action: str,
        message: str | None = None,
        config: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """Perform an intervention on a checkpointed debate."""
        data: dict[str, Any] = {"action": action}
        if message is not None:
            data["message"] = message
        if config is not None:
            data["config"] = config

        return await self._client._request(
            "POST",
            f"/api/v1/checkpoints/{checkpoint_id}/intervention",
            json=data,
        )

    # =========================================================================
    # Debate-Specific Checkpoint Operations
    # =========================================================================

    async def pause_debate(self, debate_id: str) -> dict[str, Any]:
        """Pause a debate and create a checkpoint."""
        return await self._client._request("POST", f"/api/v1/debates/{debate_id}/pause")

    # =========================================================================
    # Knowledge Mound Checkpoints
    # =========================================================================

    async def list_km(self, limit: int | None = None) -> dict[str, Any]:
        """List Knowledge Mound checkpoints."""
        params: dict[str, Any] = {}
        if limit is not None:
            params["limit"] = limit

        return await self._client._request(
            "GET",
            "/api/v1/km/checkpoints",
            params=params if params else None,
        )

    async def create_km(
        self,
        name: str,
        workspace_id: str | None = None,
    ) -> dict[str, Any]:
        """Create a Knowledge Mound checkpoint."""
        data: dict[str, Any] = {"name": name}
        if workspace_id is not None:
            data["workspace_id"] = workspace_id

        return await self._client._request("POST", "/api/v1/km/checkpoints", json=data)

    async def get_km(self, name: str) -> dict[str, Any]:
        """Get a Knowledge Mound checkpoint."""
        return await self._client._request("GET", f"/api/v1/km/checkpoints/{name}")

    async def compare_km(self, name: str, compare_to: str) -> dict[str, Any]:
        """Compare two Knowledge Mound checkpoints."""
        return await self._client._request(
            "GET",
            f"/api/v1/km/checkpoints/{name}/compare",
            params={"compare_to": compare_to},
        )

    async def compare_km_checkpoints(self, checkpoint_a: str, checkpoint_b: str) -> dict[str, Any]:
        """Compare two named Knowledge Mound checkpoints directly.

        Uses the documented POST /api/v1/km/checkpoints/compare contract
        (body: checkpoint_a, checkpoint_b).
        """
        return await self._client._request(
            "POST",
            "/api/v1/km/checkpoints/compare",
            json={"checkpoint_a": checkpoint_a, "checkpoint_b": checkpoint_b},
        )

    async def restore_km(self, name: str) -> dict[str, Any]:
        """Restore a Knowledge Mound checkpoint."""
        return await self._client._request("POST", f"/api/v1/km/checkpoints/{name}/restore")

    async def delete_km(self, name: str) -> dict[str, Any]:
        """Delete a Knowledge Mound checkpoint."""
        return await self._client._request("DELETE", f"/api/v1/km/checkpoints/{name}")

File: sdk/typescript/src/namespaces/checkpoints.ts
/**
 * Checkpoints Namespace API
 *
 * Provides debate checkpoint management for pause, resume, and intervention.
 */

export interface Checkpoint {
  id: string;
  debate_id: string;
  status: 'active' | 'resumed' | 'expired';
  round: number;
  created_at: string;
  expires_at?: string;
  metadata?: Record<string, unknown>;
}

export interface ResumableDebate {
  debate_id: string;
  checkpoint_id: string;
  task: string;
  round: number;
  paused_at: string;
}

export interface InterventionRequest {
  action: string;
  message?: string;
  config?: Record<string, unknown>;
}

export interface KMCheckpoint {
  name: string;
  workspace_id?: string;
  created_at: string;
  size_bytes?: number;
  node_count?: number;
  metadata?: Record<string, unknown>;
}

export interface CheckpointComparison {
  checkpoint_a: string;
  checkpoint_b: string;
  additions: number;
  deletions: number;
  modifications: number;
  details?: Record<string, unknown>;
}

interface CheckpointsClientInterface {
  request<T = unknown>(
    method: string,
    path: string,
    options?: { params?: Record<string, unknown>; body?: unknown }
  ): Promise<T>;
}

export class CheckpointsAPI {
  constructor(private client: CheckpointsClientInterface) {}

  // ===========================================================================
  // Debate Checkpoints
  // ===========================================================================

  /**
   * List all checkpoints.
   *
   * Supports filtering by debate via `debate_id` and by `status`
   * (documented GET /api/v1/checkpoints query params).
   */
  async list(params?: {
    limit?: number;
    offset?: number;
    debate_id?: string;
    status?: string;
  }): Promise<{ checkpoints: Checkpoint[]; total: number }> {
    return this.client.request('GET', '/api/v1/checkpoints', {
      params: params as Record<string, unknown>,
    });
  }

  /**
   * Get resumable debates with active checkpoints.
   */
  async listResumable(): Promise<{ debates: ResumableDebate[] }> {
    return this.client.request('GET', '/api/v1/checkpoints/resumable');
  }

  /**
   * Get a specific checkpoint.
   */
  async get(checkpointId: string): Promise<Checkpoint> {
    return this.client.request('GET', `/api/v1/checkpoints/${checkpointId}`);
  }

  /**
   * Resume a debate from a checkpoint.
   */
  async resume(checkpointId: string): Promise<{ debate_id: string; resumed: boolean }> {
    return this.client.request('POST', `/api/v1/checkpoints/${checkpointId}/resume`);
  }

  /**
   * Delete a checkpoint.
   */
  async delete(checkpointId: string): Promise<{ deleted: boolean }> {
    return this.client.request('DELETE', `/api/v1/checkpoints/${checkpointId}`);
  }

  /**
   * Perform an intervention on a checkpointed debate.
   */
  async intervene(
    checkpointId: string,
    body: InterventionRequest
  ): Promise<{ success: boolean; message?: string }> {
    return this.client.request('POST', `/api/v1/checkpoints/${checkpointId}/intervention`, {
      body,
    });
  }

  /**
   * Pause a debate and create a checkpoint.
   */
  async pauseDebate(debateId: string): Promise<Checkpoint> {
    return this.client.request('POST', `/api/v1/debates/${debateId}/pause`);
  }

  // ===========================================================================
  // Knowledge Mound Checkpoints
  // ===========================================================================

  /**
   * List Knowledge Mound checkpoints.
   */
  async listKM(params?: {
    limit?: number;
  }): Promise<{ checkpoints: KMCheckpoint[] }> {
    return this.client.request('GET', '/api/v1/km/checkpoints', {
      params: params as Record<string, unknown>,
    });
  }

  /**
   * Create a Knowledge Mound checkpoint.
   */
  async createKM(body: {
    name: string;
    workspace_id?: string;
  }): Promise<KMCheckpoint> {
    return this.client.request('POST', '/api/v1/km/checkpoints', { body });
  }

  /**
   * Get a Knowledge Mound checkpoint.
   */
  async getKM(name: string): Promise<KMCheckpoint> {
    return this.client.request('GET', `/api/v1/km/checkpoints/${name}`);
  }

  /**
   * Compare two Knowledge Mound checkpoints.
   */
  async compareKM(
    name: string,
    compareTo: string
  ): Promise<CheckpointComparison> {
    return this.client.request('GET', `/api/v1/km/checkpoints/${name}/compare`, {
      params: { compare_to: compareTo },
    });
  }

  /**
   * Compare two named Knowledge Mound checkpoints directly.
   *
   * Uses the documented POST /api/v1/km/checkpoints/compare contract
   * (body: checkpoint_a, checkpoint_b).
   */
  async compareKMCheckpoints(
    checkpointA: string,
    checkpointB: string
  ): Promise<CheckpointComparison> {
    return this.client.request('POST', '/api/v1/km/checkpoints/compare', {
      body: { checkpoint_a: checkpointA, checkpoint_b: checkpointB },
    });
  }

  /**
   * Restore a Knowledge Mound checkpoint.
   */
  async restoreKM(name: string): Promise<{ restored: boolean }> {
    return this.client.request('POST', `/api/v1/km/checkpoints/${name}/restore`);
  }

  /**
   * Delete a Knowledge Mound checkpoint.
   *
   * @param name - The checkpoint name to delete.
   * @returns Confirmation of deletion.
   */
  async deleteKM(name: string): Promise<{ deleted: boolean }> {
    return this.client.request('DELETE', `/api/v1/km/checkpoints/${name}`);
  }
}
